db/compaction/compaction_job.cc

```diff
@@ -353,35 +356,75 @@ void CompactionJob::GenSubcompactionBoundaries() {
   // Group the ranges into subcompactions
   const double min_file_fill_percent = 4.0 / 5;
   int base_level = v->storage_info()->base_level();
+  uint64_t max_subcompactions = static_cast<uint64_t>(c->max_subcompactions());
   uint64_t max_output_files = static_cast<uint64_t>(std::ceil(
       sum / min_file_fill_percent /
       MaxFileSizeForLevel(
           *(c->mutable_cf_options()), out_lvl,
           c->immutable_options()->compaction_style, base_level,
           c->immutable_options()->level_compaction_dynamic_level_bytes)));
-  uint64_t subcompactions =
-      std::min({static_cast<uint64_t>(ranges.size()),
-                static_cast<uint64_t>(c->max_subcompactions()),
-                max_output_files});
+  uint64_t subcompactions = std::min({static_cast<uint64_t>(ranges.size()),
+                                      max_subcompactions, max_output_files});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // The number of files in the start level cannot be larger than
+    // ranges.size() since ranges also contains the number of files from other
+    // levels Note that mutex has been locked, so it is safe to access
+    // bg_compaction_scheduled and update it
+    subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files,
+         static_cast<uint64_t>(available_compactions_ + 1)});
+    // We only reserve threads if the required subcompactions is more than the
+    // number of allowed subcompactions
+    if (max_subcompactions < subcompactions) {
+      int available_threads = env_->ReserveThreads(
+          (int)(subcompactions - max_subcompactions), thread_pri_);
+      // The reserved threads might be less than the expected one, we have to
+      // reduce the subcompactions to enforce the thread limit
+      subcompactions =
+          static_cast<uint64_t>(available_threads) + max_subcompactions;
+      if (bg_compaction_scheduled_ != nullptr) {
+        *bg_compaction_scheduled_ += available_threads;
+      }
+      reserved_threads_ = available_threads;
+    }
+  }
```

Comment:
I suggest we have better coding abstraction by using self-explanatory variable names and functions. Here are my preliminary ideas along with a few questions made obvious to me after adding coding abstraction. 

```
// Some static private helper functions
static private uint64_t DecideSubcompactionNumByLimits(uint64_t specified_max_subcompaction_limit, uint64_t compaction_input_limit, uint64_t compaction_output_limit) {
    return std::min({specified_max_subcompaction_limit, compaction_input_limit, compaction_output_limit})
}

static void AdjustSubcompactionNumByThreadsReservation(uint64_t* num_subcompactions,  uint64_t num_existing_subcompaction_threads, Env* env_, Env::Priority thread_pri,  uint64_t*  num_reserved_threads) {
    const uint64_t num_subcompaction_threads_to_reserve =(*num_subcompactions) - num_existing_subcompaction_threads;
    const uint64_t num_subcompaction_threads_reserved = env_->ReserveThreads(num_subcompaction_threads_to_reserve, thread_pri);
    *num_subcompactions = num_subcompaction_threads_reserved + num_existing_subcompaction_threads;
    *num_reserved_threads = num_subcompaction_threads_reserved
}

if (c->immutable_options()->compaction_pri == kRoundRobin) {
    // Question 1: why we use `c->num_input_files(0))` instead of `ranges.size()` for kRoundRobin? Your comment only explain why we can do that but not why we should do that
    // Question 2: We probably need a more contextual name than "available_threads" cuz the "availability" can be based on many criteria   ..... 
    uint64_t subcompactions = DecideSubcompactionNumByLimits(static_cast<uint64_t>(available_threads), static_cast<uint64_t>(c->num_input_files(0))), max_output_files);
    
    // Question 3: From our offline chat, I understood the usage of c->max_subcompactions() as to honor user-specified _mutable_db_options.max_subcompactions. 
    // But this field is only conditionally set to this value when the passed-in max_subcompactions_ == 0 (see https://github.com/facebook/rocksdb/blob/7.4.fb/db/compaction/compaction.cc#L263). 
    // Then should we still use the passed-in max_subcompactions_ like this PR does?
    AdjustSubcompactionNumByThreadsReservation(&subcompactions, c->max_subcompactions() , env_, thread_pri_,  &reserved_threads_)
  
    // Question 4: should bg_compaction_scheduled_ be charged against subcompactions OR the reserved_threads_? 
    // I thought bg_compaction_scheduled_ is related to your available_compactions_ hence it should be charged against subcompactions. But I see `*bg_compaction_scheduled_ -= released_threads;` below.
    bg_compaction_scheduled_ += subcompactions (?)
}



CompactionJob::Run() {
    ...
    // Revert changes to shared counters used above like env_'s threads by releasing, reserved_threads_, bg_compaction_scheduled_ 
    ...
}
```

Reply:
Thanks for so nice and detailed suggestion. I will work on this.